torchao/dtypes/utils.py

```diff
@@ -0,0 +1,46 @@
+from typing import Dict, Callable, Any
+from collections import defaultdict
+import functools
+
+# torch_function and torch_dispatch operator dispatch registrations
+_ATEN_OP_OR_TORCH_FN_TABLE: Dict[Callable, Dict[Any, Any]] = defaultdict(dict)
+
+def _implements(cls, aten_ops_or_torch_fns):
+    """Use this decorator to implement a function for an aten ops in __torch_dispatch__
+    (if user passed in a list of ops)
+    or torch function in __torch_function__ (if user passed in a single object)
+    """
+    if not isinstance(aten_ops_or_torch_fns, (list, tuple)):
+        aten_ops_or_torch_fns = [aten_ops_or_torch_fns]
+    def decorator(func):
+        for op in aten_ops_or_torch_fns:
+            @functools.wraps(op)
+            def wrapper(*args, **kwargs):
+                return func(*args, **kwargs)
+
+            _ATEN_OP_OR_TORCH_FN_TABLE[cls][op] = wrapper
```

Comment:
Part of me still doesn't like having the function table be shared among subclasses.... I don't have a great reason against though lol so that's not blocking

Reply:
Actually curious to hear more. I want us to get the aesthetics right

Reply:
I actually dont have any logically good reason lol it is more just gut check I dont love global things that tie a bunch of potentially disperate ideas/implementations together

The dispatcher in PyTorch is a giant global vtable that ties everything together so there are obvious counter points thats why this super nit and not blocking

Reply:
This breaks QLora in TorchTune. I plannning to add a FSDP unit test in TorchAO. Probably need to revert NF4 changes

```
[rank0]:   File "/data/users/weif/pytorch-official/pytorch/torch/distributed/_tensor/dispatch.py", line 205, in dispatch
[rank0]:     local_results = op_call(*local_tensor_args, **op_info.local_kwargs)
[rank0]:   File "/data/users/weif/pytorch-official/pytorch/torch/_ops.py", line 666, in __call__
[rank0]:     return self_._op(*args, **kwargs)
[rank0]:   File "/data/users/weif/ao/torchao/dtypes/nf4tensor.py", line 541, in __torch_function__
[rank0]:     return _ATEN_OP_OR_TORCH_FN_TABLE[cls][func](*args, **kwargs)
[rank0]:   File "/data/users/weif/ao/torchao/dtypes/utils.py", line 25, in wrapper
[rank0]:     return func(*args, **kwargs)
[rank0]: TypeError: nf4_detach() missing 1 required positional argument: 'args'
```